### ISSUE TITLE
"Remote "docker" (oci) is not an InstanceServer"

### ISSUE THREAD

**Issue by PriceChild** [lxc/terraform-provider-incus#179]
```
data "incus_image" "go-vod" {
  remote = "docker"
  name = "radialapps/go-vod"
}

resource "incus_instance" "go-vod" {
    name      = "go-vod"
    image     = data.incus_image.go-vod.fingerprint
    ephemeral = false
}
```

produces:

> ╷
│ Error: Failed to retrieve Incus ImageServer
│ 
│   with data.incus_image.go-vod,
│   on go-vod.tf line 1, in data "incus_image" "go-vod":
│    1: data "incus_image" "go-vod" {
│ 
│ Remote "docker" (oci) is not an InstanceServer
╵

using:
> $ tofu -version
> OpenTofu v1.8.6
> on linux_amd64
> + provider registry.opentofu.org/lxc/incus v0.2.0

This is following on from https://discuss.linuxcontainers.org/t/incus-tofu-keeping-oci-images-updated/22001

In short, I'm just hoping for a simpler way of keeping instances up to date with upstream image updates, other than:

```
tofu apply -replace="incus_image.go-vod"
```

**Comment by stgraber** [comment]
I thought @maveonair got this working before.

**Comment by stgraber** [comment]
@maveonair yeah, a copied OCI image with the auto-update flag on will have the image update in the background. Obviously that just updates the cached image on the machine, it doesn't affect any running instances.

**Comment by maveonair** [comment]
> @maveonair yeah, a copied OCI image with the auto-update flag on will have the image update in the background. Obviously that just updates the cached image on the machine, it doesn't affect any running instances.

Yes, as this is the default behavior for system container and VM images. 

I understand why one would want to update the application containers themselves when a new version is available, but that is definitely not the job of the Terraform provider, but rather a feature of Incus itself. Right?

**Comment by stgraber** [comment]
Yeah and even at the Incus level, this would be rather dangerous and isn't a thing that Docker does either. If you delete and create again, you'll get the new version. If you run `incus rebuild` you will also get the new version, but in either case you may lose data if you've not configured things right.

**Comment by maveonair** [comment]
> Yeah and even at the Incus level, this would be rather dangerous and isn't a thing that Docker does either. If you delete and create again, you'll get the new version. If you run `incus rebuild` you will also get the new version, but in either case you may lose data if you've not configured things right.

Yes, we know this behavior from using Docker or Kubernetes, but there even a newly started container loses its entire status in this case.

@PriceChild I'm not sure what result you expect now, as neither this provider nor Incus will probably do an automatic update and recreation of the image for you?

**Comment by stgraber** [comment]
So anyway:

```
data "incus_image" "go-vod" {
  remote = "docker"
  name = "radialapps/go-vod"
}

resource "incus_instance" "go-vod" {
    name      = "go-vod"
    image     = data.incus_image.go-vod.fingerprint
    ephemeral = false
}
```

Is invalid as even if the `incus_image` data would work with OCI (which it can't as OCI images can't be queried by fingerprint), the `incus_instance` part wouldn't work as you're referring to a fingerprint in the local image store and not pointing to a remote image.

So your best bet is to have the instance just directly created from the image you want:

```
resource "incus_instance" "go-vod" {
    name      = "go-vod"
    image     = "docker:radialapps/go-vod"
    ephemeral = false
}
```

The Incus server will cache and refresh the image, so rebuilding the instance will pick up the new image. Or you can use the image version within the `image` property as suggested by @maveonair which would then let you bump it and recreate the instance.

**Comment by PriceChild** [comment]
Thanks for all of your time on this!

> The Incus server will cache and refresh the image, so rebuilding the instance will pick up the new image.

Does this require the following?
```
tofu apply -replace="incus_instance.go-vod"
```

**Comment by PriceChild** [comment]
For what it's worth, I resolved this for myself with a module: https://github.com/PriceChild/terraform-incus-oci-image-updating

Now, whenever I `tofu apply`, the docker hub api is checked and the image replaced if necessary due to a changed fingerprint.

I would love to see support for this in the provider itself. I'm assuming I'm not missing something obvious...?

**Comment by maveonair** [comment]
I believe the auto-update of OCI images should still be handled within Incus itself rather than the provider. The provider's role is to apply the configuration changes as defined, so if a specific image version is set and later changed, it will update accordingly.

**Comment by PriceChild** [comment]
@maveonair Sure, no objection. But does that actually happen? I do not think it does.

My understanding is that if you create an image based on a tag e.g. `latest`, it is stored in incus under the fingerprint as it's name e.g. `abcd2134`. It does not store the original tag. Upstream's tag might be repointed to a new fingerprint but as the original fingerprint's contents don't change...

I would love to be corrected. Do you gave an example showing otherwise?

Thanks!

**Comment by maveonair** [comment]
That’s right. Incus doesn’t have an update mechanism right now and it does what you described. Maybe you could suggest that as a new feature?

**Comment by PriceChild** [comment]
Apologies, I think I missunderstood your "should" earlier. You meant that Incus is the right place to fix it, not here. Not that Incus already supported this.
Sorry for the confusion, thank you.

**Comment by maveonair** [comment]
> Apologies, I think I missunderstood your "should" earlier. You meant that Incus is the right place to fix it, not here. Not that Incus already supported this. Sorry for the confusion, thank you.

No problem and thanks a lot for your understanding 😊.

**Comment by victoitor** [comment]
> The data source is used to reference an existing image in the Incus server, but not to reference an OCI image from an OCI registry.

I think it makes sense to allow the incus_image data source to also reference an image from a remote. In this particular case, an OCI image from an OCI registry.

This data source seems to be mostly to query the fingerprint and those can be obtained from remote images as well. This is even used in a solution provided by @stgraber in [this forum post](https://discuss.linuxcontainers.org/t/docker-oci-parameter-passing-incus-6-3/21114/15).

Doing so should cleanly allow for an instance replace using the [`replace_triggered_by` lifecycle argument](https://opentofu.org/docs/language/resources/behavior/#replace_triggered_by).

It might not be the best production setup for reproducible builds, by keeping versions with a possible CICD pipeline, but it is very clean and easy for homelabbers.

**Comment by maveonair** [comment]
> > The data source is used to reference an existing image in the Incus server, but not to reference an OCI image from an OCI registry.
> 
> I think it makes sense to allow the incus_image data source to also reference an image from a remote. In this particular case, an OCI image from an OCI registry.
> 
> This data source seems to be mostly to query the fingerprint and those can be obtained from remote images as well. This is even used in a solution provided by [@stgraber](https://github.com/stgraber) in [this forum post](https://discuss.linuxcontainers.org/t/docker-oci-parameter-passing-incus-6-3/21114/15).
> 
> Doing so should cleanly allow for an instance replace using the [`replace_triggered_by` lifecycle argument](https://opentofu.org/docs/language/resources/behavior/#replace_triggered_by).
> 
> It might not be the best production setup for reproducible builds, by keeping versions with a possible CICD pipeline, but it is very clean and easy for homelabbers.

Please see https://github.com/lxc/terraform-provider-incus/issues/179#issuecomment-2512819942

**Comment by victoitor** [comment]
> [@stgraber](https://github.com/stgraber) is this actually implemented for OCI images too: https://linuxcontainers.org/incus/docs/main/image-handling/#auto-update?

Is it possible to set this through an incus_image resource?

**Comment by maveonair** [comment]
> > [@stgraber](https://github.com/stgraber) is this actually implemented for OCI images too: https://linuxcontainers.org/incus/docs/main/image-handling/#auto-update?
> 
> Is it possible to set this through an incus_image resource?

The auto-update only updates the image on the Incus server, but it does not re-deploy an instance. Until now, there is no auto-update image and redeploy a OCI or system container or VM.

**Comment by victoitor** [comment]
> The auto-update only updates the image on the Incus server, but it does not re-deploy an instance. Until now, there is no auto-update image and redeploy a OCI or system container or VM.

It could be configured to do so with the [`replace_triggered_by` lifecycle argument](https://opentofu.org/docs/language/resources/behavior/#replace_triggered_by). Just make it replace triggered by changing the image fingerprint.

**Comment by maveonair** [comment]
I don’t see this properly working… But what I did is that I created a small script that checks the latest SHA digest of an image and then updates the Terraform files accordingly.

**Comment by victoitor** [comment]
It is supposed to work. At least according to the documentation. 

What would be the issue? Would the incus image update not be reflected in the terraform resource?
